server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java

```diff
@@ -1396,12 +1429,50 @@ private Object bootstrapSinksFromDisk()
     return committed.getMetadata();
   }
 
+  /**
+   * Update the state of the appenderator when adding a sink.
+   *
+   * @param identifier sink identifier
+   * @param sink sink to be added
+   */
+  private void addSink(SegmentIdWithShardSpec identifier, Sink sink)
+  {
+    sinks.put(identifier, sink);
+    // Asoociate the base segment of a sink with its string identifier
+    // Needed to get the base segment using upgradedFromSegmentId of a pending segment
+    idToPendingSegment.put(identifier.asSegmentId().toString(), identifier);
+
+    // The base segment is associated with itself in the maps to maintain all the upgraded ids of a sink.
+    baseSegmentToUpgradedSegments.put(identifier, new HashSet<>());
+    baseSegmentToUpgradedSegments.get(identifier).add(identifier);
+    upgradedSegmentToBaseSegment.put(identifier, identifier);
+
+    sinkTimeline.add(
+        sink.getInterval(),
+        sink.getVersion(),
+        identifier.getShardSpec().createChunk(sink)
+    );
+  }
+
   private ListenableFuture<?> abandonSegment(
```

Comment:
Could you add a javadoc to this method? When exactly is a segment abandoned?

Reply:
Every segment is abandoned when the StreamAppenderator is closed or cleared.
A segment is also marked to be abandoned by the StreamAppenderatorDriver when it has been handed off